File: mobile/app/src/main/java/com/odyssey/travelplanner/data/AccountRepository.kt
package com.odyssey.travelplanner.data

import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.functions.functions
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.storage.storage
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.util.UUID

@Serializable
private data class UserDataRow(
    @SerialName("user_id") val userId: String,
    val key: String,
    val value: JsonObject,
)

data class AccountProfile(
    val avatarUrl: String?,
    val notificationsEnabled: Boolean,
    val language: String = "RU",
    val themePreference: ThemePreference = ThemePreference.SYSTEM,
    val tripRemindersEnabled: Boolean = true,
    val cancellationRemindersEnabled: Boolean = true,
    val reminderHour: Int = 9,
    val onboardingCompleted: Boolean = false,
    val createTripHintSeen: Boolean = false,
    val addPlaceHintSeen: Boolean = false,
    /** True when the account_profile row already existed before this read. */
    val hasStoredProfile: Boolean = false,
) {
    val darkTheme: Boolean
        get() = themePreference == ThemePreference.DARK
}

class AccountRepository(private val client: SupabaseClient) {
    suspend fun loadProfile(): AccountProfile {
        val userId = client.auth.currentUserOrNull()?.id?.toString() ?: return AccountProfile(null, false)
        val row = client.from("user_data").select {
            filter { eq("user_id", userId); eq("key", "account_profile") }
        }.decodeList<UserDataRow>().firstOrNull()
        val storedTheme = row?.value?.get("theme")?.jsonPrimitive?.contentOrNull
        val themePreference = when {
            storedTheme != null -> ThemePreference.fromStorage(storedTheme)
            row?.value?.get("dark_theme")?.jsonPrimitive?.content == "true" -> ThemePreference.DARK
            row != null -> ThemePreference.LIGHT
            else -> ThemePreference.SYSTEM
        }
        return AccountProfile(
            avatarUrl = client.resolveTripPhotoReference(row?.value?.get("avatar_url")?.jsonPrimitive?.contentOrNull),
            notificationsEnabled = row?.value?.get("notifications_enabled")?.jsonPrimitive?.content == "true",
            language = row?.value?.get("language")?.jsonPrimitive?.contentOrNull ?: "RU",
            themePreference = themePreference,
            tripRemindersEnabled = row?.value?.get("trip_reminders_enabled")?.jsonPrimitive?.content?.let { it == "true" } ?: true,
            cancellationRemindersEnabled = row?.value?.get("cancellation_reminders_enabled")?.jsonPrimitive?.content?.let { it == "true" } ?: true,
            reminderHour = row?.value?.get("reminder_hour")?.jsonPrimitive?.content?.toIntOrNull()?.coerceIn(0, 23) ?: 9,
            onboardingCompleted = row?.value?.get("onboarding_completed")?.jsonPrimitive?.content == "true",
            createTripHintSeen = row?.value?.get("create_trip_hint_seen")?.jsonPrimitive?.content == "true",
            addPlaceHintSeen = row?.value?.get("add_place_hint_seen")?.jsonPrimitive?.content == "true",
            hasStoredProfile = row != null,
        )
    }

    suspend fun updateProfile(
        avatarUrl: String?,
        notificationsEnabled: Boolean,
        language: String? = null,
        darkTheme: Boolean? = null,
        tripRemindersEnabled: Boolean? = null,
        cancellationRemindersEnabled: Boolean? = null,
        reminderHour: Int? = null,
        themePreference: ThemePreference? = null,
        onboardingCompleted: Boolean? = null,
        createTripHintSeen: Boolean? = null,
        addPlaceHintSeen: Boolean? = null,
    ) {
        val userId = client.auth.currentUserOrNull()?.id?.toString() ?: throw AuthSessionRequiredException()
        val existing = client.from("user_data").select {
            filter { eq("user_id", userId); eq("key", "account_profile") }
        }.decodeList<UserDataRow>().firstOrNull()?.value
        val value = existing?.toMutableMap() ?: mutableMapOf()
        canonicalTripPhotoReference(avatarUrl)?.let { value["avatar_url"] = JsonPrimitive(it) }
        value["notifications_enabled"] = JsonPrimitive(notificationsEnabled)
        language?.let { value["language"] = JsonPrimitive(it) }
        themePreference?.let {
            value["theme"] = JsonPrimitive(it.storageValue)
            value["dark_theme"] = JsonPrimitive(it == ThemePreference.DARK)
        } ?: darkTheme?.let {
            value["dark_theme"] = JsonPrimitive(it)
            if (value["theme"] == null) {
                value["theme"] = JsonPrimitive(if (it) ThemePreference.DARK.storageValue else ThemePreference.LIGHT.storageValue)
            }
        }
        tripRemindersEnabled?.let { value["trip_reminders_enabled"] = JsonPrimitive(it) }
        cancellationRemindersEnabled?.let { value["cancellation_reminders_enabled"] = JsonPrimitive(it) }
        reminderHour?.coerceIn(0, 23)?.let { value["reminder_hour"] = JsonPrimitive(it) }
        onboardingCompleted?.let { value["onboarding_completed"] = JsonPrimitive(it) }
        createTripHintSeen?.let { value["create_trip_hint_seen"] = JsonPrimitive(it) }
        addPlaceHintSeen?.let { value["add_place_hint_seen"] = JsonPrimitive(it) }
        client.from("user_data").upsert(
            UserDataRow(
                userId = userId,
                key = "account_profile",
                value = JsonObject(value),
            ),
        ) { onConflict = "user_id,key" }
    }

    suspend fun updateAppearance(language: String, themePreference: ThemePreference) {
        val profile = loadProfile()
        updateProfile(
            profile.avatarUrl,
            profile.notificationsEnabled,
            language = language,
            themePreference = themePreference,
        )
    }

    suspend fun updateAppearance(language: String, darkTheme: Boolean) {
        updateAppearance(
            language,
            if (darkTheme) ThemePreference.DARK else ThemePreference.LIGHT,
        )
    }

    suspend fun updateOnboardingState(
        onboardingCompleted: Boolean? = null,
        createTripHintSeen: Boolean? = null,
        addPlaceHintSeen: Boolean? = null,
    ) {
        val profile = loadProfile()
        updateProfile(
            avatarUrl = profile.avatarUrl,
            notificationsEnabled = profile.notificationsEnabled,
            language = profile.language,
            themePreference = profile.themePreference,
            tripRemindersEnabled = profile.tripRemindersEnabled,
            cancellationRemindersEnabled = profile.cancellationRemindersEnabled,
            reminderHour = profile.reminderHour,
            onboardingCompleted = onboardingCompleted,
            createTripHintSeen = createTripHintSeen,
            addPlaceHintSeen = addPlaceHintSeen,
        )
    }

    /**
     * The web app stores its first-run tutorial state in this generic user_data
     * record. Reading the same record keeps the Android and web onboarding in
     * sync without introducing another table or column.
     */
    suspend fun loadWebOnboardingCompleted(): Boolean? {
        val userId = client.auth.currentUserOrNull()?.id?.toString() ?: return null
        val row = client.from("user_data").select {
            filter { eq("user_id", userId); eq("key", "web-onboarding") }
        }.decodeList<UserDataRow>().firstOrNull() ?: return null
        return when (row.value["completed"]?.jsonPrimitive?.contentOrNull?.lowercase()) {
            "true" -> true
            "false" -> false
            else -> null
        }
    }

    suspend fun updateWebOnboardingState(completed: Boolean) {
        val userId = client.auth.currentUserOrNull()?.id?.toString() ?: throw AuthSessionRequiredException()
        client.from("user_data").upsert(
            UserDataRow(
                userId = userId,
                key = "web-onboarding",
                value = buildJsonObject {
                    put("completed", completed)
                    put("completedAt", java.time.Instant.now().toString())
                },
            ),
        ) { onConflict = "user_id,key" }
    }

    suspend fun uploadProfilePhoto(bytes: ByteArray): String {
        require(bytes.isNotEmpty()) { "Не удалось прочитать изображение" }
        val userId = client.auth.currentUserOrNull()?.id?.toString() ?: throw AuthSessionRequiredException()
        val path = "$userId/profile/${UUID.randomUUID()}.jpg"
        client.storage.from("trip-photos").upload(path, bytes)
        val reference = storedTripPhotoReference(path)
        return client.resolveTripPhotoReference(reference) ?: error("Не удалось открыть изображение")
    }

    suspend fun changePassword(password: String) {
        require(password.length >= 6) { "Пароль должен содержать минимум 6 символов" }
        client.auth.updateUser { this.password = password }
    }

    suspend fun deleteAccount() {
        if (client.auth.currentUserOrNull() == null) throw AuthSessionRequiredException()
        client.functions.invoke(
            function = "delete-account",
            body = buildJsonObject { },
        )
    }
}

File: ios/Ramingo/Networking/AccountRepository.swift
import Foundation

enum ThemePreference: String, Hashable, Sendable {
    case system
    case light
    case dark

    init(storageValue: String?) {
        switch storageValue?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "light": self = .light
        case "dark": self = .dark
        default: self = .system
        }
    }

    var storageValue: String { rawValue }
}

struct AccountProfile: Hashable, Sendable {
    let avatarReference: String?
    let notificationsEnabled: Bool
    let language: String
    let themePreference: ThemePreference
    let tripRemindersEnabled: Bool
    let cancellationRemindersEnabled: Bool
    let paymentRemindersEnabled: Bool
    let emailNotificationsEnabled: Bool
    let emailPaymentRemindersEnabled: Bool
    let emailRecipient: String?
    let reminderHour: Int
    let onboardingCompleted: Bool = false
    let createTripHintSeen: Bool = false
    let addPlaceHintSeen: Bool = false
    let hasStoredProfile: Bool = false

    init(
        avatarReference: String?,
        notificationsEnabled: Bool,
        language: String,
        themePreference: ThemePreference,
        tripRemindersEnabled: Bool,
        cancellationRemindersEnabled: Bool,
        paymentRemindersEnabled: Bool,
        emailNotificationsEnabled: Bool,
        emailPaymentRemindersEnabled: Bool,
        emailRecipient: String?,
        reminderHour: Int,
        onboardingCompleted: Bool = false,
        createTripHintSeen: Bool = false,
        addPlaceHintSeen: Bool = false,
        hasStoredProfile: Bool = false,
    ) {
        self.avatarReference = avatarReference
        self.notificationsEnabled = notificationsEnabled
        self.language = language
        self.themePreference = themePreference
        self.tripRemindersEnabled = tripRemindersEnabled
        self.cancellationRemindersEnabled = cancellationRemindersEnabled
        self.paymentRemindersEnabled = paymentRemindersEnabled
        self.emailNotificationsEnabled = emailNotificationsEnabled
        self.emailPaymentRemindersEnabled = emailPaymentRemindersEnabled
        self.emailRecipient = emailRecipient
        self.reminderHour = reminderHour
        self.onboardingCompleted = onboardingCompleted
        self.createTripHintSeen = createTripHintSeen
        self.addPlaceHintSeen = addPlaceHintSeen
        self.hasStoredProfile = hasStoredProfile
    }

    var darkTheme: Bool { themePreference == .dark }

    static let defaults = AccountProfile(
        avatarReference: nil,
        notificationsEnabled: false,
        language: "RU",
        themePreference: .system,
        tripRemindersEnabled: true,
        cancellationRemindersEnabled: true,
        paymentRemindersEnabled: true,
        emailNotificationsEnabled: true,
        emailPaymentRemindersEnabled: true,
        emailRecipient: nil,
        reminderHour: 9,
        onboardingCompleted: false,
        createTripHintSeen: false,
        addPlaceHintSeen: false,
        hasStoredProfile: false,
    )
}

private struct UserDataRow: Decodable {
    let userID: String
    let key: String
    let value: JSONValue

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case key
        case value
    }
}

private struct UserDataUpsert: Encodable {
    let userID: String
    let key: String
    let value: JSONValue

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case key
        case value
    }
}

final class AccountRepository {
    private let client: SupabaseClient

    init(client: SupabaseClient) {
        self.client = client
    }

    func loadProfile() async throws -> AccountProfile {
        guard let userID = client.currentUser?.id else { return .defaults }
        let safeUserID = userID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userID
        let path = "rest/v1/user_data?select=user_id,key,value&user_id=eq.\(safeUserID)&key=eq.account_profile"
        let rows: [UserDataRow] = try await client.request(
            path,
            method: "GET",
            body: nil as EmptyAccountBody?,
            authenticated: true,
        )
        guard let value = rows.first?.value.objectValue else { return .defaults }
        let themePreference: ThemePreference
        if let storedTheme = value.text("theme").nonEmpty {
            themePreference = ThemePreference(storageValue: storedTheme)
        } else if value.flag("dark_theme") == true {
            themePreference = .dark
        } else {
            // Android treats an existing legacy profile without a theme as light.
            themePreference = .light
        }
        return AccountProfile(
            avatarReference: value.text("avatar_url").nonEmpty,
            notificationsEnabled: value.flag("notifications_enabled") ?? false,
            language: value.text("language", fallback: "RU").uppercased(),
            themePreference: themePreference,
            tripRemindersEnabled: value.flag("trip_reminders_enabled") ?? true,
            cancellationRemindersEnabled: value.flag("cancellation_reminders_enabled") ?? true,
            paymentRemindersEnabled: value.flag("payment_reminders_enabled") ?? true,
            emailNotificationsEnabled: value.flag("email_notifications_enabled") ?? true,
            emailPaymentRemindersEnabled: value.flag("email_payment_reminders_enabled") ?? true,
            emailRecipient: value.text("email_recipient").nonEmpty.flatMap(normalizeNotificationEmail),
            reminderHour: min(max(value.integer("reminder_hour") ?? 9, 0), 23),
            onboardingCompleted: value.flag("onboarding_completed") ?? false,
            createTripHintSeen: value.flag("create_trip_hint_seen") ?? false,
            addPlaceHintSeen: value.flag("add_place_hint_seen") ?? false,
            hasStoredProfile: true,
        )
    }

    func updateProfile(
        avatarReference: String?,
        notificationsEnabled: Bool,
        language: String? = nil,
        darkTheme: Bool? = nil,
        themePreference: ThemePreference? = nil,
        tripRemindersEnabled: Bool? = nil,
        cancellationRemindersEnabled: Bool? = nil,
        paymentRemindersEnabled: Bool? = nil,
        emailNotificationsEnabled: Bool? = nil,
        emailPaymentRemindersEnabled: Bool? = nil,
        emailRecipient: String? = nil,
        reminderHour: Int? = nil,
        onboardingCompleted: Bool? = nil,
        createTripHintSeen: Bool? = nil,
        addPlaceHintSeen: Bool? = nil,
    ) async throws {
        guard let userID = client.currentUser?.id else { throw SupabaseClientError.cancelled }
        let existing = try await loadProfileValue(userID: userID)
        var next = existing
        if let avatarReference { next["avatar_url"] = .string(avatarReference) }
        next["notifications_enabled"] = .boolean(notificationsEnabled)
        language.map { $0.uppercased() }.map { next["language"] = .string($0) }
        if let themePreference {
            next["theme"] = .string(themePreference.storageValue)
            next["dark_theme"] = .boolean(themePreference == .dark)
        } else if let darkTheme {
            next["dark_theme"] = .boolean(darkTheme)
            if next["theme"] == nil {
                next["theme"] = .string(darkTheme ? ThemePreference.dark.storageValue : ThemePreference.light.storageValue)
            }
        }
        tripRemindersEnabled.map { next["trip_reminders_enabled"] = .boolean($0) }
        cancellationRemindersEnabled.map { next["cancellation_reminders_enabled"] = .boolean($0) }
        paymentRemindersEnabled.map { next["payment_reminders_enabled"] = .boolean($0) }
        emailNotificationsEnabled.map { next["email_notifications_enabled"] = .boolean($0) }
        emailPaymentRemindersEnabled.map { next["email_payment_reminders_enabled"] = .boolean($0) }
        if let emailRecipient {
            guard let normalized = normalizeNotificationEmail(emailRecipient) else {
                throw SupabaseClientError.invalidInput("Укажите корректный e-mail")
            }
            next["email_recipient"] = .string(normalized)
        }
        reminderHour.map { next["reminder_hour"] = .number(Double(min(max($0, 0), 23))) }
        onboardingCompleted.map { next["onboarding_completed"] = .boolean($0) }
        createTripHintSeen.map { next["create_trip_hint_seen"] = .boolean($0) }
        addPlaceHintSeen.map { next["add_place_hint_seen"] = .boolean($0) }
        let _: [UserDataRow] = try await client.request(
            "rest/v1/user_data?on_conflict=user_id,key",
            method: "POST",
            body: [UserDataUpsert(userID: userID, key: "account_profile", value: .object(next))],
            authenticated: true,
            extraHeaders: ["Prefer": "resolution=merge-duplicates,return=representation"],
        )
    }

    func uploadProfilePhoto(data: Data) async throws -> String {
        guard !data.isEmpty else { throw SupabaseClientError.server(status: 422, message: "Не удалось прочитать изображение.") }
        guard let userID = client.currentUser?.id else { throw SupabaseClientError.cancelled }
        let path = "\(userID)/profile/\(UUID().uuidString.lowercased()).jpg"
        try await client.uploadStorageObject(path: path, data: data)
        let reference = "storage://trip-photos/\(path)"
        do {
            let profile = try await loadProfile()
            try await updateProfile(
                avatarReference: reference,
                notificationsEnabled: profile.notificationsEnabled,
                language: profile.language,
                themePreference: profile.themePreference,
                tripRemindersEnabled: profile.tripRemindersEnabled,
                cancellationRemindersEnabled: profile.cancellationRemindersEnabled,
                paymentRemindersEnabled: profile.paymentRemindersEnabled,
                emailNotificationsEnabled: profile.emailNotificationsEnabled,
                emailPaymentRemindersEnabled: profile.emailPaymentRemindersEnabled,
                emailRecipient: profile.emailRecipient,
                reminderHour: profile.reminderHour,
                onboardingCompleted: profile.onboardingCompleted,
                createTripHintSeen: profile.createTripHintSeen,
                addPlaceHintSeen: profile.addPlaceHintSeen,
            )
            if let previousReference = profile.avatarReference, previousReference != reference {
                await client.deleteStorageReference(previousReference)
            }
        } catch {
            try? await client.deleteStorageObject(path: path)
            throw error
        }
        return reference
    }

    func changePassword(_ password: String) async throws {
        try await client.updatePassword(password)
    }

    func deleteAccount() async throws {
        guard client.currentUser != nil else { throw SupabaseClientError.cancelled }
        try await client.invokeFunction("delete-account", body: EmptyAccountBody())
    }

    func loadWebOnboardingCompleted() async throws -> Bool? {
        guard let userID = client.currentUser?.id else { return nil }
        let safeUserID = userID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userID
        let path = "rest/v1/user_data?select=user_id,key,value&user_id=eq.\(safeUserID)&key=eq.web-onboarding"
        let rows: [UserDataRow] = try await client.request(
            path,
            method: "GET",
            body: nil as EmptyAccountBody?,
            authenticated: true,
        )
        guard let value = rows.first?.value.objectValue else { return nil }
        return value.flag("completed")
    }

    func updateWebOnboardingState(completed: Bool) async throws {
        guard let userID = client.currentUser?.id else { throw SupabaseClientError.cancelled }
        let value: JSONValue = .object([
            "completed": .boolean(completed),
            "completedAt": .string(ISO8601DateFormatter().string(from: Date())),
        ])
        let _: [UserDataRow] = try await client.request(
            "rest/v1/user_data?on_conflict=user_id,key",
            method: "POST",
            body: [UserDataUpsert(userID: userID, key: "web-onboarding", value: value)],
            authenticated: true,
            extraHeaders: ["Prefer": "resolution=merge-duplicates,return=representation"],
        )
    }

    private func loadProfileValue(userID: String) async throws -> [String: JSONValue] {
        let safeUserID = userID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? userID
        let path = "rest/v1/user_data?select=user_id,key,value&user_id=eq.\(safeUserID)&key=eq.account_profile"
        let rows: [UserDataRow] = try await client.request(
            path,
            method: "GET",
            body: nil as EmptyAccountBody?,
            authenticated: true,
        )
        return rows.first?.value.objectValue ?? [:]
    }
}

private let notificationEmailPattern = try! Regex(#"^[^\s@]+@[^\s@]+\.[^\s@]+$"#)

func normalizeNotificationEmail(_ value: String) -> String? {
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard normalized.count <= 254, normalized.wholeMatch(of: notificationEmailPattern) != nil else { return nil }
    return normalized
}

private struct EmptyAccountBody: Encodable {}
